### ISSUE TITLE
org.apache.cassandra.io.FSWriteError: java.nio.file.AccessDeniedException

### ISSUE THREAD

**Issue by dejwsz** [Stratio/cassandra-lucene-index#57]
Hi,
today such error happens on one of my test nodes where lucene plugin is added: 
/var/lib/cassandra/lucene is the directory where the lucene index is stored. After that error cassandra stops responding and I needed to restart cassandra service.

ERROR [MigrationStage:1] 2015-10-29 12:25:10,206 CassandraDaemon.java:227 - Exception in thread Thread[MigrationStage:1,5,main]
org.apache.cassandra.io.FSWriteError: java.nio.file.AccessDeniedException: /var/lib/cassandra/lucene
        at org.apache.cassandra.io.util.FileUtils.deleteWithConfirm(FileUtils.java:135) ~[apache-cassandra-2.1.11.jar:2.1.11]
        at org.apache.cassandra.io.util.FileUtils.deleteRecursive(FileUtils.java:381) ~[apache-cassandra-2.1.11.jar:2.1.11]
        at com.stratio.cassandra.lucene.service.LuceneIndex.delete(LuceneIndex.java:213) ~[cassandra-lucene-index-plugin-2.2.3.0-SNAPSHOT.jar:na]
        at com.stratio.cassandra.lucene.service.RowService.delete(RowService.java:223) ~[cassandra-lucene-index-plugin-2.2.3.0-SNAPSHOT.jar:na]
        at com.stratio.cassandra.lucene.Index.removeIndex(Index.java:201) ~[cassandra-lucene-index-plugin-2.2.3.0-SNAPSHOT.jar:na]
        at com.stratio.cassandra.lucene.Index.removeIndex(Index.java:180) ~[cassandra-lucene-index-plugin-2.2.3.0-SNAPSHOT.jar:na]
        at org.apache.cassandra.db.index.SecondaryIndexManager.removeIndexedColumn(SecondaryIndexManager.java:264) ~[apache-cassandra-2.1.11.jar:2.1.11]
        at org.apache.cassandra.db.index.SecondaryIndexManager.reload(SecondaryIndexManager.java:130) ~[apache-cassandra-2.1.11.jar:2.1.11]
        at org.apache.cassandra.db.ColumnFamilyStore.reload(ColumnFamilyStore.java:200) ~[apache-cassandra-2.1.11.jar:2.1.11]
        at org.apache.cassandra.db.DefsTables.updateColumnFamily(DefsTables.java:427) ~[apache-cassandra-2.1.11.jar:2.1.11]
        at org.apache.cassandra.db.DefsTables.mergeColumnFamilies(DefsTables.java:295) ~[apache-cassandra-2.1.11.jar:2.1.11]
        at org.apache.cassandra.db.DefsTables.mergeSchemaInternal(DefsTables.java:194) ~[apache-cassandra-2.1.11.jar:2.1.11]
        at org.apache.cassandra.db.DefsTables.mergeSchema(DefsTables.java:166) ~[apache-cassandra-2.1.11.jar:2.1.11]
        at org.apache.cassandra.service.MigrationManager$2.runMayThrow(MigrationManager.java:394) ~[apache-cassandra-2.1.11.jar:2.1.11]
        at org.apache.cassandra.utils.WrappedRunnable.run(WrappedRunnable.java:28) ~[apache-cassandra-2.1.11.jar:2.1.11]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471) ~[na:1.7.0_91]
        at java.util.concurrent.FutureTask.run(FutureTask.java:262) ~[na:1.7.0_91]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145) ~[na:1.7.0_91]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615) [na:1.7.0_91]
        at java.lang.Thread.run(Thread.java:745) [na:1.7.0_91]
Caused by: java.nio.file.AccessDeniedException: /var/lib/cassandra/lucene
        at sun.nio.fs.UnixException.translateToIOException(UnixException.java:84) ~[na:1.7.0_91]
        at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102) ~[na:1.7.0_91]
        at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107) ~[na:1.7.0_91]
        at sun.nio.fs.UnixFileSystemProvider.implDelete(UnixFileSystemProvider.java:244) ~[na:1.7.0_91]
        at sun.nio.fs.AbstractFileSystemProvider.delete(AbstractFileSystemProvider.java:103) ~[na:1.7.0_91]
        at java.nio.file.Files.delete(Files.java:1079) ~[na:1.7.0_91]
        at org.apache.cassandra.io.util.FileUtils.deleteWithConfirm(FileUtils.java:131) ~[apache-cassandra-2.1.11.jar:2.1.11]
        ... 19 common frames omitted
ERROR [SharedPool-Worker-1] 2015-10-29 12:25:10,207 QueryMessage.java:135 - Unexpected error during query
java.lang.RuntimeException: java.util.concurrent.ExecutionException: FSWriteError in /var/lib/cassandra/lucene
        at org.apache.cassandra.utils.FBUtilities.waitOnFuture(FBUtilities.java:376) ~[apache-cassandra-2.1.11.jar:2.1.11]
        at org.apache.cassandra.service.MigrationManager.announce(MigrationManager.java:375) ~[apache-cassandra-2.1.11.jar:2.1.11]
        at org.apache.cassandra.service.MigrationManager.announceColumnFamilyUpdate(MigrationManager.java:296) ~[apache-cassandra-2.1.11.jar:2.1.11]
        at org.apache.cassandra.cql3.statements.DropIndexStatement.announceMigration(DropIndexStatement.java:100) ~[apache-cassandra-2.1.11.jar:2.1.11]
        at org.apache.cassandra.cql3.statements.DropIndexStatement.execute(DropIndexStatement.java:88) ~[apache-cassandra-2.1.11.jar:2.1.11]
        at org.apache.cassandra.cql3.QueryProcessor.processStatement(QueryProcessor.java:238) ~[apache-cassandra-2.1.11.jar:2.1.11]
        at com.stratio.cassandra.lucene.IndexQueryHandler.process(IndexQueryHandler.java:141) ~[cassandra-lucene-index-plugin-2.2.3.0-SNAPSHOT.jar:na]
        at org.apache.cassandra.transport.messages.QueryMessage.execute(QueryMessage.java:122) ~[apache-cassandra-2.1.11.jar:2.1.11]
        at org.apache.cassandra.transport.Message$Dispatcher.channelRead0(Message.java:439) [apache-cassandra-2.1.11.jar:2.1.11]
        at org.apache.cassandra.transport.Message$Dispatcher.channelRead0(Message.java:335) [apache-cassandra-2.1.11.jar:2.1.11]
        at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:105) [netty-all-4.0.23.Final.jar:4.0.23.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:333) [netty-all-4.0.23.Final.jar:4.0.23.Final]
        at io.netty.channel.AbstractChannelHandlerContext.access$700(AbstractChannelHandlerContext.java:32) [netty-all-4.0.23.Final.jar:4.0.23.Final]
        at io.netty.channel.AbstractChannelHandlerContext$8.run(AbstractChannelHandlerContext.java:324) [netty-all-4.0.23.Final.jar:4.0.23.Final]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471) [na:1.7.0_91]
        at org.apache.cassandra.concurrent.AbstractTracingAwareExecutorService$FutureTask.run(AbstractTracingAwareExecutorService.java:164) [apache-cassandra-2.1.11.jar:2.1.11]
        at org.apache.cassandra.concurrent.SEPWorker.run(SEPWorker.java:105) [apache-cassandra-2.1.11.jar:2.1.11]
        at java.lang.Thread.run(Thread.java:745) [na:1.7.0_91]
Caused by: java.util.concurrent.ExecutionException: FSWriteError in /var/lib/cassandra/lucene
        at java.util.concurrent.FutureTask.report(FutureTask.java:122) ~[na:1.7.0_91]
        at java.util.concurrent.FutureTask.get(FutureTask.java:188) ~[na:1.7.0_91]
        at org.apache.cassandra.utils.FBUtilities.waitOnFuture(FBUtilities.java:372) ~[apache-cassandra-2.1.11.jar:2.1.11]
        ... 17 common frames omitted

**Comment by dejwsz** [comment]
This happened after I tried to remove the index which had been created before. Right after the drop command in cqlsh I received this error. A moment later the whole cluster began to have problems. Only restart could help.

**Comment by dejwsz** [comment]
Hmm, after restart I created the index once again. Everything went OK but the data are not written to index directory, on every node I have only two files: segments_1 and write_lock. Index seems to be empty while the column family holds some data.

**Comment by dejwsz** [comment]
Finally this helped me to repair the state of the lucene index:
nodetool invalidaterowcache
nodetool invalidatekeycache
nodetool rebuild_index epodb notes notes_index

Everything else ended up with the FSWriteError. Index is working well again now.

**Comment by ealonsodb** [comment]
First of all, 
I can see in the logs that you are using index version 2.2.3.0 with cassandra 2.1.11. That's not a good idea. Please use index version 2.1.11.0 with cassandra 2.1.11. 

The FSWriteError suggests that the OS user running cassandra does not have permisions to write to that folder. Have you changed the version of the plugin before deleting the index??

**Comment by dejwsz** [comment]
Yes I've used master trunk at first because at that time there wasn't version for 2.1.11 of cassandra - but it worked very well for a some time until I decided to remove the index - what is interesting cassandra user had read/write/owner rights to this directory - index worked, data were indexed and I were able to query it. After the issue arose I changed the version of plugin and then tried and finally managed to fix it as I wrote earlier. I can't explain why cassandra claimed it couldn't delete the index - the permissions were OK in the file system.
